### DISCUSSION TITLE
Minimum cursor sensitivity

### DISCUSSION THREAD

**Discussion by someoneyouspite** [ppy/osu#25045] [Q&A]
### Type

Game behaviour

### Bug description

Cursor sensitivity should allow values under 0.1 for users with high DPI tracking devices.
I primarily use a 25600 DPI with a windows sensitivity multiplier of 0.0625 to represent the feel of 1600 DPI at a default sensitivity, this is an inconvenience when I have to switch DPI because of the clamped sensitivity value.
Could explicit definition within the settings be allowed while retaining clamped values for the slider, or at least respect the setting in config file without clamping it unless the interface is adjusted?

### Version

2023.924.1-lazer

### Logs

[performance-draw.log](https://github.com/ppy/osu/files/12838864/performance-draw.log)
[performance-input.log](https://github.com/ppy/osu/files/12838865/performance-input.log)
[performance-update.log](https://github.com/ppy/osu/files/12838866/performance-update.log)
[runtime.log](https://github.com/ppy/osu/files/12838867/runtime.log)
[updater.log](https://github.com/ppy/osu/files/12838868/updater.log)
[database.log](https://github.com/ppy/osu/files/12838869/database.log)
[input.log](https://github.com/ppy/osu/files/12838870/input.log)
[legacy-ipc.log](https://github.com/ppy/osu/files/12838871/legacy-ipc.log)
[network.log](https://github.com/ppy/osu/files/12838872/network.log)
[performance.log](https://github.com/ppy/osu/files/12838873/performance.log)
[performance-audio.log](https://github.com/ppy/osu/files/12838874/performance-audio.log)

**Comment by peppy** [comment]
Turn down your DPI instead. What the hell..

**Comment by someoneyouspite** [comment]
I figured disabling raw input works as a hack for now. But just because I get negligible benefits for latency beyond 3200 dpi doesn't mean I can't be a bit crazy in trying to provide more data per poll.
A comparison would be how people want as many frames as possible even though it's 8x or more than their monitor's refresh rate.

**Comment by someoneyouspite** [comment]
If you want a drastic example of why I fear low DPI, take 1 DPI moving 10 inches in 1 second, you'd get at least 100ms of latency per dot not counting everything else. Sure at 3200 DPI we start dealing in the sub ms range for most movements but latency is the thief of joy, and I enjoy the placebo I get from having even 1ns less of it.

**Comment by peppy** [comment]
> you'd get at least 100ms of latency per dot not counting everything else

that's not how it works......

**Reply by someoneyouspite** [reply]
I'm sorry you disagree, but you can easily test this for yourself to notice how lower DPI introduces "pixel skipping" which would be comparable to frame drops not matching the refresh rate. You lose information with lower DPI, just because poll rate does report the exact same position every 1hz doesn't mean there isn't perceivable latency from my arm movement to the movement of the cursor on the screen from DPI. Plenty of independent benchmarks have been done to showcase these results.
If I'm misunderstanding something, I'd love for you to elaborate on how it actually works.

**Reply by someoneyouspite** [reply]
https://i.imgur.com/jbMTq06.gif
Here's an example of this "perceivable latency" I'm talking about. In reality I'm moving my arm with a smooth motion, however the data from the hardware only has so many steps it can provide to match the "sensitivity feel" of a higher DPI. If it has less dots than poll rate over a given measure of time you are wasting the performance on poll rate, and with a low enough DPI to match the sensitivity you want, a single dot could end up being larger than a pixel size.
In the case of osu this isn't very beneficial due to the precision being lost on the pixel grid, but in the case of a 3d game environment that listens to every dot to rotate a camera with radians it is quite beneficial at reducing the feel of latency and preventing drift from low DPI errors adding up.

**Reply by someoneyouspite** [reply]
I'm gonna be honest, it's a bit frustrating that most new gaming mice are releasing with 25-30k DPIs and you think that people wouldn't want to push their devices to the limit like they do with every other one. Disregarding legitimate benefits from high DPI that you can easily verify on google from innumerable sources and test yourself, just to deny changing a sensitivity value in the configuration because you disagree with how something works, it's a bit petty to say the least.

**Comment by someoneyouspite** [comment]
that's how it works...... (at least with my hardware)
having a high enough sensitivity with a low enough DPI will leave larger gaps in between positions and add latency (maybe not to the cpu) but latency where it **REALLY MATTERS** from my motor control input to the visual feedback bleeding from the pixels on screen

**Comment by Omena0** [comment]
I would also like this change, my current sensitivity 0.21 is too low and 0.22 is too high. I'd want to be able to at least set it to 0.215 in the config, but the config parser rounds it to the nearest 1/100th.

**Reply by peppy** [reply]
If you drop your mouse DPI by 50% you will be able to choose the value you want (double it) with no other discernible change.

**Reply by someoneyouspite** [reply]
It is genuinely baffling that years later, your only response to users dealing with this is still just "change your DPI."

We are talking about changing a single, hardcoded minimum clamp value for a float variable. That's it. A 30-second UI or config parser tweak.

Instead of just saying "I don't want to do this" or engaging in an actual technical discussion, your default reaction is to be petty, insult your users' intelligence, and casually try to gaslight them by saying "that's not how it works." It is how it works. Spatial quantization error and polling rate saturation on modern 4K+ Hz mice are mathematically measurable phenomena. Just because you personally choose to ignore the physics of sensor input latency doesn't mean the data is wrong, and it certainly doesn't justify talking down to the people playing your game.

This sheer arrogance is exactly why I canceled my osu! supporter subscription. It’s not even about the sensitivity slider anymore; it’s about the principle of financially supporting a developer who actively dismisses, belittles, and mocks the community over incredibly simple, quality-of-life feature requests.

Keep the artificial clamp if it protects your ego, but don't expect people to keep funding this kind of hostility when they just want to use their hardware properly.

**Reply by peppy** [reply]
The argument here is to adjust precision, not clamping.

> Spatial quantization error and polling rate saturation on modern 4K+ Hz mice are mathematically measurable phenomena.

DO you have testing covering this? I'd be interested in seeing how this can actually affect a normal user, apart from creating more issues. One of the reason we have these limits in place is that dealing with such high polling and low floating point values is that they **create new issues** which we are avoiding.

I'm not intending to be arrogant. I'm intending to make sure we don't shoot ourselves in the foot at the whim of ~~hardware~~ marketing companies that just want to inflate their numbers on paper so people choose them over competition.

**Reply by someoneyouspite** [reply]
I appreciate the response, but respectfully, the argument is about clamping, which is why the title of this discussion is 'Minimum cursor sensitivity' and my original post explicitly asked to 'allow values under 0.1' and 'retaining clamped values for the slider, or at least respect the setting in config file without clamping it.' Precision is a secondary issue brought up by another user two years later, the core request has always been about the hard-coded minimum limit.

Regarding the request for testing and evidence: it feels like moving the goalposts to demand mathematically measured studies now, especially when my initial explanations were met with a dismissive 'that's not how it works' without any technical elaboration or evidence provided on your end for over two years. There is extensive, publicly available documentation and testing from sensor engineers and independent reviewers (like Optimum Tech, Battle(Non)Sense, and various mouse sensor whitepapers) demonstrating how high DPI combined with lower in-game multipliers minimizes input latency and avoids pixel skipping. The burden of proof shouldn't suddenly fall on the user to write a whitepaper just to justify an unlocked config value.

> One of the reason we have these limits in place is that dealing with such high polling and low floating point values is that they create new issues which we are avoiding.

Could you elaborate on what specific issues are caused by allowing a float multiplier lower than 0.1 in the config parser? Modern environments handle standard 32-bit floats perfectly fine well below those values.

If you don't want to implement it because you feel high DPI is just 'hardware marketing,' that is your prerogative as the lead developer. But asking users to fundamentally change their global hardware settings just to accommodate an arbitrary UI clamp in a single game, rather than just allowing a lower float threshold, still feels like an unnecessary roadblock.

**Comment by shin0kaze** [comment]
I wish it will be implemented some day. cuz now, osu can't handle mouses above 36k dpi... It's not that hard to implement, just allow to write any float value above 0 near the slider. Yes, I can lower dpi, but this ruin whole point of having high dpi, and changing each time it just for a single game a bit... awkward maybe?
